fusion/README.md

```diff
@@ -0,0 +1,198 @@
+                                              InstPtrListType &)>;
+
+
+    static void defaultFusionOpr(Fusion & inst, InstPtrListType & in,
+                                 InstPtrListType & out)
+    {
+        out.insert(out.end(), in.begin(), in.end());
+        in.clear();
+    }
+
+```
+
+# Usage of the draft PR
+
+Pull the DPR, build and run the local testbench, create the doxygen docs.
+
+The dpr is on branch jeffnye-gh/fusion_dpr
+
+```
+cd fusion
+mkdir -p release; cd release
+cmake ..
+make -j32 regress
+make docs
+```
+
+# C++ API and testbench
+
+A C++ API is defined and a baseline set of implementations are provided. 
+The implementations are linked to the performance model using the conventional 
+callback mechanism. The C++ API is intended for compiled execution, regressions, etc.
+
+The testbench is contained in TestBench, test/testbench.h.
+
+## C++ API summary
+
+
+ The primary classes in the API are:
+
+```
+    Fusion
+    FusionGroup (FusionGroupCfg)
+    MachineInfo
+    FieldExtractor
+```
+
+The top level class is Fusion, fusion/fusion.h.  This is the primary
+interface to Sparta based decoders.
+
+FusionGroup holds the definition of a fusable group of instructions. FusionGroup
+provides functor handles for user defined methods. FusionGroupCfg is a
+ctor helper struct/class for FusionGroup.
+
+MachineInfo is a placeholder (currently) for machine microarchitecture information pertinent to fusion and fusion performance modeling.
+
+FieldExtractor is a shim between FusionGroup functors and Mavis. It provides
+boolean and field extraction methods. I expect this to change, either remove
+the shim or increase the number of FieldExtractor setter/getter methods.
+
+There are multiple input mechanisms intended to simplify implementing
+the fusion operation. Currently I am using UID. There will be support
+for opcodes, assembly text, a DSL syntax and possibly a Json input format.
+
+UID based fusion is the example in this draft. The others are planned.
+
+## Other classes
+
+The support classes for Fusion are:
+
+```
+  FusionContext         fusioncontext.h
+  FusionExceptionBase   fusionexceptions.h
+  using and typedefs    fusiontypes.h
+  RadixTrie             radixtrie.h
+```
+
+There is a single context supported in this version of the implementation but
+generalization to multiple contexts has placeholders in the source.
+
+FusionExceptionBase is the parent for a a number of fusion/dsl/json/run-time exceptions specific to fusion.
+
+RadixTrie is experimental. It is implemented and provided for performance analysis during development. The currently used container is an unordered\_map 
+used for functionality and test development.
+
+## Usage within a Sparta unit
+
+The currently considered procedure for Fusion is an implementation in a 
+Sparta unit typically in the Decode. The primary input to the 
+fusionOperator() is a form of the InstQueue, sparta::Queue<InstPtr>.
```

Comment:
I would guess that the operator is templatized (unless a lambda) that can take any type of iterable.